components/layout/wrapper.rs

```diff
@@ -774,13 +774,41 @@ pub trait ThreadSafeLayoutNode : Clone + Copy + Sized + PartialEq {
                 // Precompute non-eagerly-cascaded pseudo-element styles if not
                 // cached before.
                 let style_pseudo = other.style_pseudo_element();
-                if !style_pseudo.is_eagerly_cascaded() &&
-                   !self.borrow_layout_data().unwrap().style_data.per_pseudo.contains_key(&style_pseudo) {
-                    let mut data = self.mutate_layout_data().unwrap();
-                    let new_style = context.stylist
-                                           .computed_values_for_pseudo(&style_pseudo,
-                                                                       data.style_data.style.as_ref());
-                    data.style_data.per_pseudo.insert(style_pseudo.clone(), new_style.unwrap());
+                match style_pseudo.cascade_type() {
+                    // Already computed during the cascade.
+                    PseudoElementCascadeType::Eager => {},
+                    PseudoElementCascadeType::Precomputed => {
+                        if !self.borrow_layout_data()
+                                .unwrap().style_data
+                                .per_pseudo.contains_key(&style_pseudo) {
+                            let mut data = self.mutate_layout_data().unwrap();
+                            let new_style =
+                                context.stylist
+                                       .precomputed_values_for_pseudo(&style_pseudo,
+                                                                      data.style_data.style.as_ref());
+                            data.style_data.per_pseudo
+                                .insert(style_pseudo.clone(), new_style.unwrap());
+                        }
+                    }
+                    PseudoElementCascadeType::Lazy => {
+                        panic!("Lazy pseudo-elements can't be used in Servo \
+                               since accessing the DOM tree during layout \
+                               could be unsafe.")
```

Comment:
We could still use a subset of selector-matching though (attributes + tag name at least), enough for avoiding doing the whole cascade for some private pseudos, like `::-servo-details-summary`, which only needs access to the `details` tag and the `open` attribute.

Would this be desirable (with appropriate docs of course), or is it just a lot of overhead? I think it can be worth it but I'd like to know your opinion.